skynet/modules/ttt/rag/web_crawler/recursive_url_loader.py

```diff
@@ -0,0 +1,112 @@
+import asyncio
+import io
+from typing import List, Optional, Set
+
+import aiohttp
+import requests
+from fake_useragent import UserAgent
+
+from langchain_community.document_loaders.recursive_url_loader import RecursiveUrlLoader
+from langchain_core.documents import Document
+from langchain_core.utils.html import extract_sub_links
+from pypdf import PdfReader
+
+from skynet.logs import get_logger
+
+ua = UserAgent()
+log = get_logger(__name__)
+
+
+class SkynetRecursiveUrlLoader(RecursiveUrlLoader):
+    async def _async_get_child_links_recursive(
+        self,
+        url: str,
+        visited: Set[str],
+        *,
+        session: Optional[aiohttp.ClientSession] = None,
+        depth: int = 0,
+    ) -> List[Document]:
+        if depth >= self.max_depth:
+            return []
+
+        # Disable SSL verification because websites may have invalid SSL certificates,
```

Comment:
Smart!

Reply:
can't take credit for this, I just monkey patched the original langchain method to add pdf parsing support, I'd gladly take another approach...